packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,121 @@
+
+For example, consider a table of geographic information that included location, population, flag, and possibly other location-related information. The location column could be either a continent, country, state, or any sub-regional construct. It would be reasonable to want to present this information in a way that allowed a user to collapse/expand any outer regional grouping, which itself would contain all rows whose locations existed inside it (e.g. 'United States' would be parent of all rows whose location was a U.S. state).
+
+### Lazy Loading
+
+In addition to just supporting data hierarchy with a fully provided set of data, we also need to support the scenario where a row is presented as a parent, but none of its children are currently represented in the data. Upon expanding the row, the client is then expected to load all of the data for that parent and provide it to the table (via the standard `setData` method).
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#890: Design for hierarchical data support in tables](https://github.com/ni/nimble/issues/890)
+-   [#861: Table hierarchical data support](https://github.com/ni/nimble/issues/861)
+-   [Prototype using flat list that includes a parentId](https://github.com/ni/nimble/tree/data-hierarchy-flat-list-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-qzwoshcidz.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+-   [Prototype with user data being hierarchical](https://github.com/ni/nimble/tree/data-hierarchy-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-yncupvnoes.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+
+## Implementation / Design
+
+Users will provide hierarchical data as a flat list (just as they do for non-hierarchical data), but will have to include a field in their data that is representative of the "parent Id" for that record of data.
+
+### `Table` API
+
+```ts
+public Table() {
+    ...
+    // This attribute determines which field in the record specifies the id that is the
+    // parent row. If it is not set, the row has no parent.
+    @attr({ attribute: 'parent-id-field-name' })
+    public parentIdFieldName?: string;
+
+    // This field would be an option for how we allow a user to specify the field
+    // name where they will provide a value indicating whether or not that row
+    // is a parent row. See the below discussion on lazy-loading. If it is not set,
+    // the row is presumed child-less (and thus will show no expand/collapse button).
+    @attr({ attribute: 'has-children-field-name' })
+    public hasChildrenFieldName?: string;
+
+    // This attribute will control the behavior of what happens when a user selects a parent
+    // row.
+    @attr({ attribute: 'parent-selects-children'})
+    public parentSelectsChildren?: boolean;
```

Comment:
We should be sure to think through what the behavior is with all combinations of `parentSelectsChildren` + `selectionMode`. I think the particularly unclear case might be single-selection + parentSelectsChildren=true. My initial thought is that parent rows wouldn't be selectable with single-selection, but we should make sure that makes sense and  probably include a short section in this HLD to describe the expected behavior in each case.

Reply:
I'll leave this thread open for discussion, but after chatting with @leslieab, the set of valid behaviors does _not_ include one where `selectionMode` is `single` _and_ a parent can select its children. So a possible alternative to this API is one where we provide a new option for `selectionMode` that allows multi-select but does _not_ allow a parent to select its children. Perhaps `multiWithNoAutoChildSelect`? Otherwise, we will have an API that allows for an invalid configuration, that we would likely just have to document what the expected behavior would be.

Reply:
I had a similar thought before seeing this thread. My first idea was to add a new selection mode: `selectionMode = 'none' | 'single' | 'multiple' | 'multiple-with-children'` or something like that.

Another idea: `selectionMode` is ignored for hierarchical tables and we add a new attribute `hierarchySelectionMode = 'none' | 'single' | 'multiple' | 'multiple-with-children' | ...`? 

I could imagine wanting other modes in future like `'single-leaf' | 'multiple-leaf'`.

I think all of these options open the door to invalid combinations and we just need to live with that and rely on our existing patterns to report it or ignore it. With the `parentSelectsChildren` attribute it'd be invalid/ignored if it's set when `selectionMode = 'none'` or `selectionMode = 'single'`. With the new selection modes it'd be invalid if the table isn't hierarchical but a client sets a hierarchical selection mode.

My initial vote would be to add new mode(s) to the existing `selectionMode` property.

Reply:
This may be pedantic, but I wouldn't consider a `selectionMode` set to something that semantically is aligned with hierarchical data invalid when the table currently isn't displaying hierarchical data. It would simply be not used to its full extent. Whereas, having the two separate APIs of `parentSelectsChildren` (or whatever name we want for that) along with `selectionMode` _does_ enable an invalid state.

Ultimately, I think we are in violent agreement that we should have a single API, and just add a new mode to `selectionMode`.

Reply:
I guess the question front of us is do we want a mode that allows multiple selection _AND_ where selecting a parent doesn't auto-select children. If so, that might be semantically hard to capture in a new mode (I mean we can, but it might be fairly verbose).

Reply:
> I guess the question front of us is do we want a mode that allows multiple selection AND where selecting a parent doesn't auto-select children.

Why wouldn't we want that mode? It seems reasonable to me to be able to select multiple items in a hierarchical table and have it select each row individually regardless of where they exist in the hierarchy. Am I missing something? Wouldn't that just be the `'multiple'` mode? And then `'multiple-with-children'` has the different behavior when selecting a row with children?

Reply:
I think you are right. Not sure how I tripped myself up in my head.